### ISSUE TITLE
Remote Actor?

### ISSUE THREAD

**Issue by oliverzy** [RuedigerMoeller/kontraktor#7]
Hi Author,

First of all, thanks for creating such great project.

Since I cannot find the roadmap in the project, I am just curious that do you want to support the remote actor like Akka?

**Comment by RuedigerMoeller** [comment]
I am currently finishing the implementation of basics (Futures, yield()) in the trunk. As kontraktor is used in active development in a consulting engagement I have, I first want to ensure its well rounded.
Regarding remoting: I already have Fast-Cast which basically implements a very high performance actor alike remote platform (each topic is similar to an actor). 
I will later on provide a plug so you can remote Actors using fast-cast behind the scenes.
My issue with "transparent" remoting is, that some mechanics of local actors are quite expensive to provide for remote actors, so I think over-abstracting here will lead to poor performance (e.g. fast-cast can handle millions of remote calls per second, compare that to akka).
Vice versa, (multicast-) remoting has some unique features like one-to many remote calls which are not part of the actor model. But one-to many multicast is a key feature for (loss-less) fault tolerance, that's why akka (tried?) to move from TCP-style remoting to zeroMQ multicast messaging. Unfortunately they have to deal with binary libs then, so kontraktor will be at advantage (pure low-level [off-heap] java).
The result will be a platform which will not be able to remote completely transparent (as this does not yield good results), but it will be very easy to remote existing local actors with minor changes.
Its the same as with threads/actor scheduling: kontraktor is less automatic here in turn giving more control to the developer which yields in higher performance most of the time :-)

last but not least I want to customize Netty to provide an actor-style websocket webserver. Additionally FST 2.0 will enable easy serialization from/to JS+Dart.

My envisioned end-stack is angualr.js, angular.dart <=> [automatic fst-serialization] <=> netty-actored <=> kontraktor/fast-cast cluster. No xml, no j2ee :-)

a lot to do ;)

**Comment by oliverzy** [comment]
Sounds great! Do you ever consider using Vert.x core as HTTP/TCP/WebSocket(SockJS) server? It also uses Netty async IO as low-level transport component, it has a Node.js like API.

This may save your time on this part.

In terms of serialization,does fast-serialization is Java only format? I see you end-stack including angularjs, I suppose it will have a JS version as well, won't it?

**Comment by RuedigerMoeller** [comment]
I peeked at vert.x, but not in depth. The point is, netty already does all of the low level work, its very easy to get a http+websocket server running and "actorize" it on top. Netty's front API is already very node.js'y, so I see no benefit to squeeze vert.x inbetween. I already hacked a prototype and it took only some hours, so a polished version will not be much effort.

fst serialization is API compatible but  not binary compatible to jdk serialization. 1.x is pure java2java. 
I have a java-to-js serialization in the 2.0 trunk of fst, however I am not satisfied for now. JS is quite a different beast as there is no type information at runtime. So serialization JS=>java feels messy currently, Java=>JS is fine .. 
As fst 2.0 splits into a databinding and a codec (=binary format) layer, I consider adding a JSon based codec in additiona to the new cross-platform binary codec. 
I could use Jackson for that. Its pretty fast and offers some other formats like smile and bson out of the box ...
Another option would be the generation of JS/Dart classes from a given set of Java classes, however hum .. I am not sure if this is that handy in reality ..

https://github.com/RuedigerMoeller/fast-serialization/blob/mix/testshell/src/main/javascript/app.js

see the MinBin.xx commands on how JS=>Java serialization looks currently. At the bottom of the file there is a prototype how a generated Java-mirror class could look like ..
